In [29]:
import numpy as np
import pandas as pd

In [30]:
import warnings

warnings.filterwarnings('ignore')

In [31]:
from sklearn.datasets import load_boston

In [32]:
boston = load_boston()

In [33]:
boston.keys()

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename'])

In [34]:
data = boston['data']

data

array([[6.3200e-03, 1.8000e+01, 2.3100e+00, ..., 1.5300e+01, 3.9690e+02,
        4.9800e+00],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9690e+02,
        9.1400e+00],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9283e+02,
        4.0300e+00],
       ...,
       [6.0760e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
        5.6400e+00],
       [1.0959e-01, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9345e+02,
        6.4800e+00],
       [4.7410e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
        7.8800e+00]])

In [35]:
feature_names = boston['feature_names']

feature_names

array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
       'TAX', 'PTRATIO', 'B', 'LSTAT'], dtype='<U7')

In [36]:
target = boston['target']

target[:10]

array([24. , 21.6, 34.7, 33.4, 36.2, 28.7, 22.9, 27.1, 16.5, 18.9])

In [37]:
X = pd.DataFrame(data, columns = feature_names)

X.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

In [38]:
y = pd.DataFrame(target, columns=["price"])

y.head()

price
0   24.0
1   21.6
2   34.7
3   33.4
4   36.2

In [39]:
from sklearn.model_selection import train_test_split

In [40]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 42)

In [43]:
from sklearn.linear_model import LogisticRegression

In [44]:
from sklearn import preprocessing
from sklearn import utils

In [50]:
lab_enc = preprocessing.LabelEncoder()
y_train2 = lab_enc.fit_transform(y_train)
print(y_train2)
print(utils.multiclass.type_of_target(y_train))
print(utils.multiclass.type_of_target(y_train.astype('int')))
print(utils.multiclass.type_of_target(y_train2))

[142  93  75 111  73 128 166   0 145  69  97 111 108  91 193  40  53 135
  83  80  91 182 112  86  67 144 175 123  20  34  23  61 163 133  31  38
 194 100  81 118  59  25   1 153 130  76  54  34 109  45 136 173 109 124
 128 194 169 156 121 101  36 185  75 158 131  98  97   7 191 186 155  14
  54  82 165  61 194  87 112  28  78 108 142 151 109  99 119 160 122  95
 125   9 131 111  43  10  75 119 126  80 118  30 143 135 168  30  47  24
  40  16 127  57   6  94  47 113 157 181 148  87 159 184 122  88 100  64
  43   2  83  94 140 148  90 110  37  19 179  56  13 110 107  85  97 194
   8  70 178  49  52 102  88  32 192 118 107  58 149 172 183  65 100  68
 189  21  69  50 100   3  86  34  27  66 111  93 111 115 194 132 102 194
   7 113  97  71  66  58  31  22 132  97 140  87 100  35  17 147 132  14
 112 123 190  99   5 174 188  61 136 180  36 139  12  73 187 137 128  48
  53  29 194 102 161  44  41  34 122 167 103 194  11  30 102  63  62 128
  52 110  83 162 127  64  28 169  12  81 137 113 17

In [51]:
lr = LogisticRegression()

In [52]:
lr.fit(X_train, y_train2)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=None, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)

In [49]:
y_pred = lr.predict(X_test)

y_pred

array([147, 194,  44, 111, 112, 100,  76,  30,  60,  77,  97, 135,   0,
        97, 102,  81,  87,  34, 194,  54,  92, 100,  33, 100,  54,  73,
       111,  36,  97,  75,  70, 111,  94,  99, 112,  47, 142, 102,  97,
       111,  75, 140, 188, 100, 100,  54,  33, 111, 112, 160, 135, 170,
        76,  86, 194, 128,  54, 179, 100, 102, 102, 173, 191, 111,  68,
        53,  54, 100, 169,  72,  97, 128,   3,  64,  97,   3,  86, 188,
         3,   6, 100,  34,  83,  34,  78,  97,  42, 111, 163,  50, 106,
         7,  66,  59,  88,  74, 194,   5,  36,   3,  75,  97,   3, 111,
       100,   5, 111, 111, 102, 109,   5,   3, 102, 178, 156,   7, 159,
        53,  75, 127,  87, 130,  15, 108, 126, 111,  83, 159, 118, 190,
        57, 177,  53,  70,  34, 128,  72, 188, 100,  53,  91, 127,  83,
        97,   4, 100,  31,  54,  53, 189,  36,  54], dtype=int64)

In [65]:
X_pred = lr.predict(X_train)

X_pred

array([142,  75, 100, 111,  73, 128, 166,   0,  92,  69, 111, 111, 108,
        91, 194,  54,  53, 135,  44,  77, 111, 194, 112,  77, 108, 144,
       175, 100,  34,  34,  34,  61, 163, 133,  36,  38, 194, 100,  81,
       118,  75,  94,   3, 153, 127,  77,  54,  20, 100,  45,  97, 173,
       100, 102, 118, 194, 169, 156, 121, 101,  36, 185,  75, 158, 127,
        98,  55,  14, 191, 194, 155,  14, 112, 111, 165,  50, 194,  87,
       127,   3, 102, 108, 142, 151, 135, 194, 119, 160, 182,  95, 125,
        30,  33, 100, 135,  10,  86, 119, 126,  97, 118,  34, 143, 135,
       168,   3,  47,  30,  34,  16, 127,  57,   6,  75,  40, 113, 194,
       181, 148,  87, 159, 184, 102,  81, 112,  64,  43,   2,  64,  85,
       140, 106,  12,  87,  50,  19, 179,  56,  13,  73, 107, 135,  85,
       194,   8,  70, 178, 111,  52, 102,  88,  36, 192,  96, 107,  50,
       149, 172, 183,  70,  85,  68, 189,  21,  69,  50, 100,   3,  77,
        34,  54,  50, 111, 100,  75, 115, 194, 100, 102, 178,  3

In [56]:
from sklearn.metrics import r2_score

In [67]:
r2_score(y_pred, y_test)

-1.664770725485496

#### ЗАДАНИЕ 2

In [68]:
from sklearn.ensemble import RandomForestRegressor

In [74]:
model = RandomForestRegressor(max_depth=12, random_state=42, n_estimators=1000, bootstrap=True, criterion='mse')

model

RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=12,
           max_features='auto', max_leaf_nodes=None,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=1000, n_jobs=None,
           oob_score=False, random_state=42, verbose=0, warm_start=False)

In [75]:
model.fit(X_train, y_train.values[:,0])

RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=12,
           max_features='auto', max_leaf_nodes=None,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=1000, n_jobs=None,
           oob_score=False, random_state=42, verbose=0, warm_start=False)

In [91]:
y_pred2 = model.predict(X_test)

In [92]:
r2_score(y_pred2, y_test)

0.8481499145965063

Модель RandomForestRegressor более эффективна, чем LogisticRegression, т.к. r2_score RandomForestRegressor = 0.8481499145965063, т.е. более приближен к 1 и является положительным